gangplank/spec/clouds.go

```diff
@@ -0,0 +1,263 @@
+	Bucket  string   `yaml:"bucket,omitempty" json:"bucket,omitempty"`
+	Enabled bool     `yaml:"enabled,omitempty" json:"enabled,omitempty"`
+	Public  bool     `yaml:"public,omitempty" json:"public,omitempty"`
+	Regions []string `yaml:"regions,omitempty" json:"regions,omitempty"`
+}
+
+// GetPublishCommand returns the cosa upload command for Aliyun
+func (a *Aliyun) GetPublishCommand(buildID string) (string, error) {
+	if buildID == "" {
+		return "", fmt.Errorf("No build provided")
+	}
+
+	if !a.Enabled {
+		return "", nil
+	}
+
+	if a.Public {
+		return "", fmt.Errorf("Public is not supported on Aliyun")
+	}
+
+	baseCmd := "coreos-assembler buildextend-aliyun"
+	args := []string{"--upload",
+		fmt.Sprintf("--build=%s", buildID),
+		fmt.Sprintf("--bucket=s3://%s", a.Bucket)}
+
+	if len(a.Regions) > 0 {
+		args = append(args, fmt.Sprintf("--region=%s", a.Regions[0]))
+	}
+
+	cmd := fmt.Sprintf("%s %s", baseCmd, strings.Join(args, " "))
+	return cmd, nil
+}
+
+// Aws describes the upload options for AWS images
+//  AmiPath: the bucket patch for pushing the AMI name
+//  Public: when true, mark as public
+//  GrantUser: users to grant access to ami
+//  GrantUserSnapshot: users to grant access to snapshot
+//  Regions: name of AWS regions to push to.
+type Aws struct {
+	Enabled           bool     `yaml:"enabled,omitempty" json:"enabled,omitempty"`
+	AmiPath           string   `yaml:"ami_path,omitempty" json:"ami_path,omitempty"`
+	GrantUser         []string `yaml:"grant_user,omitempty" json:"grant_user,omitempty"`
+	GrantUserSnapshot []string `yaml:"grant_user_snapshot,omitempty" json:grant_user_snapshot,omitempty"`
+	Public            bool     `yaml:"public,omitempty" json:"public,omitempty"`
+	Regions           []string `yaml:"regions,omitempty" json:"regions,omitempty"`
+}
+
+// GetPublishCommand returns the cosa upload command for Aws
+func (a *Aws) GetPublishCommand(buildID string) (string, error) {
+	if buildID == "" {
+		return "", fmt.Errorf("No build provided")
+	}
+
+	if !a.Enabled {
+		return "", nil
+	}
+
+	baseCmd := "coreos-assembler buildextend-aws"
+	args := []string{"--upload",
+		fmt.Sprintf("--build=%s", buildID),
+		fmt.Sprintf("--bucket=s3://%s", a.AmiPath)}
+
+	if len(a.Regions) > 0 {
+		args = append(args, fmt.Sprintf("--region=%s", a.Regions[0]))
+	}
+
+	if len(a.GrantUser) > 0 {
+		args = append(args, fmt.Sprintf("--grant-user %s", strings.Join(a.GrantUser, ",")))
+	}
+
+	if len(a.GrantUserSnapshot) > 0 {
+		args = append(args, fmt.Sprintf("--grant-user-snapshot %s", strings.Join(a.GrantUserSnapshot, ",")))
+	}
+
+	cmd := fmt.Sprintf("%s %s", baseCmd, strings.Join(args, " "))
+	return cmd, nil
+}
+
+// Aws describes the upload options for AWS images
+//  AmiPath: the bucket patch for pushing the AMI name
+//  Public: when true, mark as public
+//  GrantUser: users to grant access to ami
+//  GrantUserSnapshot: users to grant access to snapshot
+//  Regions: name of AWS regions to push to.
+type AwsCn struct {
```

Comment:
In general, I think it better to avoid adding duplicate structs that implement 100% the same logic. 

WDYT of changing `CloudCfgs` struct to:
```
type CloudsCfgs struct {
	Aliyun     Aliyun `yaml:"aliyun,omitempty" json:"aliyun,omitempty"`
	Aws        Aws    `yaml:"aws,omitempty" json:"aws,omitempty"`
	AwsCN      Aws    `yaml:"aws_cn,omitempty" json:"aws_cn,omitempty"`
	Azure      Azure  `yaml:"azure,omitempty" json:"azure,omitempty"`
	Gcp        Gcp    `yaml:"gcp,omitempty" json:"gcp,omitempty"`
}
```

And then change the struct for AWS to:
```
type Aws struct {
	Enabled           bool     `yaml:"enabled,omitempty" json:"enabled,omitempty"`
	AmiPath           string   `yaml:"ami_path,omitempty" json:"ami_path,omitempty"`
	GrantUser         []string `yaml:"grant_user,omitempty" json:"grant_user,omitempty"`
	GrantUserSnapshot []string `yaml:"grant_user_snapshot,omitempty" json:grant_user_snapshot,omitempty"`
	Public            bool     `yaml:"public,omitempty" json:"public,omitempty"`
	Regions           []string `yaml:"regions,omitempty" json:"regions,omitempty"`
        SecretAnnotation         string   `yaml:"secret_annotation,omitempty" json:"secret_annotation,omitempty"`
}
```

All secrets are written to `/srv/.<GANGPLANK SECRET ANNOTATION>`, i.e. `/srv/.{aws,aws-cn}/config`. The logic to form the command would be exactly the same regardless and we would have a single function for handling the logic of AWS. 

Then the definition of a China region  would look like:
```
aws_cn: 
   ...
   secret_annotation: aws-cn
aws:
   ....
   secret_annotation: aws
```

And if we need to add another sovereign region for AWS is a logic-free PR.